test/distributed/fsdp/test_fsdp_tp_integration.py

```diff
@@ -0,0 +1,285 @@
+        return torch.cat(all_grads).contiguous()
+
+    # Sync grads across ranks by using FSDP style average of gradients on all ranks.
+    def _sync_tp_module_grads(self, m, tp_pg, params_fsdp_flat_order):
+        fsdp_world_size = int(self.world_size // tp_pg.size())
+        # TP handles gradients differently from FSDP. We need to divide by tp_pg size.
+        for p in m.parameters():
+            all_params = [torch.zeros_like(p) for _ in range(fsdp_world_size)]
+            splits = tuple(params_fsdp_flat_order.values())
+            all_params = torch.cat(all_params).contiguous().split(splits)
+            for idx, key in enumerate(params_fsdp_flat_order.keys()):
+                if key not in OPS_NOT_SHARD:
+                    all_params[idx][:] = 1
+            all_params = torch.cat(all_params).contiguous().type(torch.BoolTensor)
+            cur_param = all_params.chunk(fsdp_world_size)[self.rank // tp_pg.size()]
+            # We want to sync the layer 3 to make it same as FSDP only case.
+            p_grad_device = p.grad.device
+            p_grad = p.grad.clone().detach()
+            p_grad = p_grad.cuda(self.rank)
+            dist.all_reduce(p_grad, op=dist.ReduceOp.SUM, group=tp_pg)
+            p_grad = p_grad.to(p_grad_device)
+            p.grad[~cur_param] = p_grad[~cur_param]
+            # Sharded Tensor add up all gradients, so we need to do average.
+            p.grad /= tp_pg.size()
+
+    @skip_if_lt_x_gpu(4)
+    @parametrize("model_parallel_size", [2, 4])
+    @parametrize(
+        "cpu_offload",
+        [CPUOffload(offload_params=True), CPUOffload(offload_params=False)],
+    )
+    def test_fsdp_tp_integration(self, model_parallel_size, cpu_offload):
+        """Test FSDP with TP Integration."""
+        inp_size = [2, 3, 5]
+        # Use same seed so that each rank get the same model params.
+        torch.manual_seed(0)
+        model = SimpleModel().cuda(self.rank)
+        torch.manual_seed(0)
+        model_tp = SimpleModel().cuda(self.rank)
+        params_fsdp_flat_order, sharding_info = self._params_fsdp_flat_order(
+            model, SHARD_PARAMS, model_parallel_size
+        )
+
+        # Create Input
+        input_seed = self.rank
+        torch.manual_seed(input_seed + 1)
+        input = torch.rand(*inp_size).cuda(self.rank)
+        self.assertEqual(model(input), model_tp(input))
+
+        # Generate sub-process group for TP and FSDP.
+        tp_pg, fsdp_pg = self._generate_sub_pg(model_parallel_size)
+
+        # Wrap the control group model with FSDP only.
+        model = FSDP(model, process_group=self.process_group, cpu_offload=cpu_offload)
+        sharding_specs = _generate_chunk_sharding_spec(tp_pg.size(), tp_pg)
+        # Shard the module first and then wrap with FSDP.
+        sharding_plan = _module_sharding_plan(sharding_specs)
+        shard_module(model_tp, sharding_plan, process_group=tp_pg)
```

Comment:
When rebased on `viable/strict`, if I print `model_tp.named_parameters()` after the `shard_module()` call and before wrapping with FSDP, I only see
```
net1.bias torch.Size([8]) <class 'torch.nn.parameter.Parameter'>
net2.bias torch.Size([4]) <class 'torch.nn.parameter.Parameter'>
net3.weight torch.Size([12, 4]) <class 'torch.nn.parameter.Parameter'>
net3.bias torch.Size([12]) <class 'torch.nn.parameter.Parameter'>
```
on rank 0. It seems that `net1.weight` and `net2.weight` are missing, which I believe to be the intended `ShardedTensor`s (based on `_module_sharding_plan()`).

It seems that they are not being made visible to `nn.Module` methods like `named_parameters()`. This leads to an error like:
```
Traceback (most recent call last):
  File "/fsx/users/andgu/work/pytorch/torch/testing/_internal/common_distributed.py", line 622, in run_test
    getattr(self, test_name)()
  File "/fsx/users/andgu/work/pytorch/torch/testing/_internal/common_distributed.py", line 503, in wrapper
    fn()
  File "/fsx/users/andgu/work/pytorch/torch/testing/_internal/common_utils.py", line 216, in instantiated_test
    test(self, **param_kwargs)
  File "/fsx/users/andgu/work/pytorch/torch/testing/_internal/common_distributed.py", line 145, in wrapper
    return func(*args, **kwargs)
  File "/fsx/users/andgu/work/pytorch/test/distributed/fsdp/test_fsdp_tp_integration.py", line 251, in test_fsdp_tp_integration
    self._sync_tp_module_grads(
  File "/fsx/users/andgu/work/pytorch/test/distributed/fsdp/test_fsdp_tp_integration.py", line 179, in _sync_tp_module_grads
    all_params = torch.cat(all_params).contiguous().split(splits)
  File "/fsx/users/andgu/work/pytorch/torch/_tensor.py", line 775, in split
    return super(Tensor, self).split_with_sizes(split_size, dim)
RuntimeError: start (48) + length (48) exceeds dimension size (72).
```
where `72 = 8 + 4 + 12*4 + 12`.

cc: @wanchaol

Reply:
This is indeed wired. Let me ask Wanchao about this.